storey/flow.py

```diff
@@ -1435,3 +1436,182 @@ def get_table(self, key):
+            print(f"initializing runnable {runnable.name}")
+            runnable.init()
+            if runnable.execution_mechanism == "multiprocessing":
+                num_processes += 1
+            elif runnable.execution_mechanism == "threading":
+                num_threads += 1
+            elif runnable.execution_mechanism not in ("asyncio", "naive"):
+                raise ValueError(f"Unsupported execution mechanism: {runnable.execution_mechanism}")
+
+        # enforce max
+        if self.max_processes:
+            num_processes = min(num_processes, self.max_processes)
+        if self.max_threads:
+            num_threads = min(num_threads, self.max_threads)
+
+        self._executors = {}
+        if num_processes:
+            mp_context = multiprocessing.get_context("spawn")
+            self._executors["multiprocessing"] = ProcessPoolExecutor(max_workers=num_processes, mp_context=mp_context)
+        if num_threads:
+            self._executors["threading"] = ThreadPoolExecutor(max_workers=num_threads)
+
+    async def _do(self, event):
+        if event is _termination_obj:
+            return await self._do_downstream(_termination_obj)
+        else:
+            runnables = self.select_runnables(event)
+            if runnables is None:
+                runnables = self.runnables
+            futures = []
+            runnables_encountered = set()
+            for runnable in runnables:
+                if isinstance(runnable, str):
+                    runnable = self._runnable_by_name[runnable]
+                if id(runnable) in runnables_encountered:
+                    raise ValueError(f"select_runnables() returned more than one outlet named '{runnable.name}'")
+                input = event.body if runnable.execution_mechanism == "multiprocessing" else copy.deepcopy(event.body)
+                runnables_encountered.add(id(runnable))
+                if runnable.execution_mechanism == "asyncio":
+                    future = asyncio.get_running_loop().create_task(runnable._async_run(input, event.path))
+                elif runnable.execution_mechanism == "naive":
+                    future = asyncio.get_running_loop().create_future()
+                    future.set_result(runnable._run(input, event.path))
+                else:
+                    executor = self._executors[runnable.execution_mechanism]
+                    future = asyncio.get_running_loop().run_in_executor(
+                        executor,
+                        runnable._run,
+                        input,
+                        event.path,
+                    )
+                futures.append(future)
+            results: list[_ParallelExecutionRunnableResult] = await asyncio.gather(*futures)
+            event.body = {"input": event.body, "results": {}}
+            for result in results:
+                event.body["results"][result.runnable_name] = {"runtime": result.runtime, "output": result.data}
```

Comment:
I think we should avoid modifying the original event directly and instead add all relevant information under a new key (self.name represent the name of the current step(ParallelExecution)):
```suggestion
            event.body["result"][self.name] = {"input": input}
            for result in results:
                event.body["results"][self.name][result.runnable_name] = {"runtime": result.runtime, "output": result.data}
```

Reply:
The current output is in line with the planning we've done so far, but we've discussed the fact that we should run this by DS and consider different common scenarios, and maybe need to change this format accordingly.

Therefore, leaving as is in this PR as discussed.